src/Symfony/Component/Serializer/Tests/SerializerTest.php

```diff
@@ -558,47 +558,51 @@ public function testNormalizeScalarArray()
 
     public function testDeserializeScalar()
     {
-        $serializer = new Serializer([], ['json' => new JsonEncoder()]);
+        $serializer = new Serializer([new ScalarDenormalizer()], ['json' => new JsonEncoder()]);
 
         $this->assertSame(42, $serializer->deserialize('42', 'int', 'json'));
         $this->assertTrue($serializer->deserialize('true', 'bool', 'json'));
         $this->assertSame(3.14, $serializer->deserialize('3.14', 'float', 'json'));
         $this->assertSame(3.14, $serializer->deserialize('31.4e-1', 'float', 'json'));
+        $this->assertSame(3.0, $serializer->deserialize('3', 'float', 'json')); // '3' === json_encode(3.0)
         $this->assertSame('  spaces  ', $serializer->deserialize('"  spaces  "', 'string', 'json'));
         $this->assertSame('@Ca$e%', $serializer->deserialize('"@Ca$e%"', 'string', 'json'));
     }
 
     public function testDeserializeLegacyScalarType()
     {
-        $this->expectException(LogicException::class);
-        $serializer = new Serializer([], ['json' => new JsonEncoder()]);
+        $this->expectException(NotNormalizableValueException::class);
+        $serializer = new Serializer([new ScalarDenormalizer()], ['json' => new JsonEncoder()]);
```

Comment:
These tests demonstrate that now you need to pass a `ScalarDenormalizer` instance to the `Serializer` to be able to deserialize scalar values. Shouldn't this be considered a BC break and require a compatibility layer?

Reply:
Indeed, I will resolve this conversation since it a duplicate of https://github.com/symfony/symfony/pull/39140#discussion_r528387266